### ISSUE TITLE
An API to prevent against false assumptions

### ISSUE THREAD

**Issue by adpaco-aws** [model-checking/kani#2693]
Requested feature: We know our users are afraid of assuming conditions which are `false`, and in fact some APIs have been added to prevent against this (e.g., `kani::cover!(cond)` statements). However, it would be interesting to have an API for assuming which protects against false assumptions by default.

Use case:

User writes a harness such as:
```rust
#[kani::proof]
fn dummy() {
    let x: u32 = kani::any();
    kani::assume(x > 3 && x < 1);
    [...]
}
```
User will likely be confused, since he/she/they have assumed a condition that evaluates to `false`. At some point, the user may realize this because all checks from there are unreachable. Let's suppose we have a `safe_assume` which first checks the condition and then assumes it:
```rust
#[kani::proof]
fn dummy() {
    let x: u32 = kani::any();
    kani::safe_assume(x > 3 && x < 1);
    [...]
}
```
The implementation would be straightforward: `safe_assume` would become an assertion followed by an assumption. The check coming from the assertion will be a failure, allowing the user to realize the mistake.

Link to relevant documentation (Rust reference, Nomicon, RFC): N/A

Test case:

```rust
<code>
```

**Comment by celinval** [comment]
I can see how using `kani::assume()` can be misleading for newcomers, but I don't think creating a new API that behaves exactly like an `assert!()` is the solution. I think this would likely add more confusion than solve the initial problem.

Here are a few suggestions on how we could improve this situation:
- Add a coverage check for things that follow an assumption. To raise awareness about assumptions that always evaluate to false.
- Improve API documentation. Add an example how `assume!(false)` stops the analysis, or how `assume!(cond)` translates to `if (!cond) { exit 0}`.
- Add more details to our tutorials.
- Rename `kani::assume()` to something different.

**Comment by adpaco-aws** [comment]
Right, I've given it some thought and I don't think a new API would solve this particular problem. Here are my thoughts on your suggestions:

> - Add a coverage check for things that follow an assumption. To raise awareness about assumptions that always evaluate to false.

Do you mean add a `cover` statement? My feeling is that this would not solve the problem entirely. Since our checks are not sorted in any way, the check corresponding to the `cover` statement may not be easily found.

> - Improve API documentation. Add an example how `assume!(false)` stops the analysis, or how `assume!(cond)` translates to `if (!cond) { exit 0}`.
> - Add more details to our tutorials.

This could help.

> - Rename `kani::assume()` to something different.

This is good idea, but I can't think of a better name for it. It isn't a bad time to reconsider the current name though.

To be honest, I've given some thought to this over the weekend and I'm starting to consider this as a better alternative:
 * Keep `kani::assume(cond)` (modulo the renaming discussed earlier) but add a check to prevent false assumptions.
 * Add another API that behaves like the current `kani::assume(cond)` but doesn't check anything, allowing users to constrain executions based on a certain condition. To be honest, it's not clear to me that "constraining executions based on a certain condition" is an actual use-case; maybe not taking a condition is simpler. In other words, I can see a user writing something like `kani::block()` (equivalent to current `kani::assume(false)` to block the analysis at a certain point), but I cannot see the reason to use a condition for it.
